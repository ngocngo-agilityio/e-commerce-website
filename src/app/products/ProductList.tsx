// Apis
import { getProductList } from '@apis';

// Constants
import { SORT_BY } from '@constants';

// Components
import { ProductList } from '@components';

interface Props {
  searchValue?: string;
  sortDirection?: string;
}

const Products = async ({
  searchValue,
  sortDirection,
}: Props): Promise<JSX.Element> => {
  const queryConfig = {
    name: searchValue,
    sortDirection,
    sortBy: sortDirection && SORT_BY.NAME,
  };

  // Fetch data for product list
  const { data } = await getProductList(queryConfig);

  return <ProductList data={data} />;
};

export default Products;
