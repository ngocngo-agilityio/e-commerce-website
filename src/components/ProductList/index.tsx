// Libs
import { memo } from 'react';
import isEqual from 'react-fast-compare';
import { Text, Grid, GridItem } from '@chakra-ui/react';

// Types
import { Product } from '@types';

// Components
import { ProductCard } from '@components';

interface Props {
  data: Product[];
}

const ProductList = ({ data = [] }: Props): JSX.Element => {
  return (
    <>
      {data.length > 0 ? (
        <Grid
          templateColumns={{ base: 'repeat(1, 1fr)', md: 'repeat(4, 1fr)' }}
          rowGap={{ base: '20px', md: '48px' }}
          columnGap="20px"
        >
          {data.map((product) => {
            const { id, name = '', image = '', price = 0 } = product || {};

            return (
              <GridItem key={id}>
                <ProductCard
                  key={id}
                  id={id}
                  name={name}
                  image={image}
                  price={price}
                />
              </GridItem>
            );
          })}
        </Grid>
      ) : (
        <Text textAlign="center">No products in list</Text>
      )}
    </>
  );
};

export default memo(ProductList, isEqual);
