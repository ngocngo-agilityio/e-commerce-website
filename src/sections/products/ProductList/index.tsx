// Libs
import dynamic from 'next/dynamic';
import { Stack } from '@chakra-ui/react';

// Apis
import { getProductList } from '@apis';

// Constants
import { PRODUCT_PER_PAGE, SORT_BY } from '@constants';

// Components
import { ProductList } from '@components';

const Pagination = dynamic(() => import('@components/Pagination'));

interface Props {
  searchValue?: string;
  sortDirection?: string;
  categoryIds?: string[];
  currentPage?: string;
}

const Products = async ({
  searchValue,
  sortDirection,
  categoryIds,
  currentPage,
}: Props): Promise<JSX.Element> => {
  const queryConfig = {
    name: searchValue,
    sortDirection,
    sortBy: sortDirection && SORT_BY.NAME,
    categoryIds: categoryIds,
    page: currentPage,
    limit: PRODUCT_PER_PAGE,
  };

  // Fetch data for product list
  const { data: productList, total: productsTotal } =
    await getProductList(queryConfig);

  const pagesQuantity = Math.ceil(productsTotal / PRODUCT_PER_PAGE);

  return (
    <Stack spacing="50px">
      <ProductList data={productList} />
      {productsTotal > 0 && <Pagination pagesQuantity={pagesQuantity} />}
    </Stack>
  );
};

export default Products;
