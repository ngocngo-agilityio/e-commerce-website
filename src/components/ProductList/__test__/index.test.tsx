// Libs
import { render, screen } from '@testing-library/react';

// Components
import ProductList from '..';

// Mocks
import { MOCK_PRODUCT_LIST } from '@mocks';

describe('ProductList component', () => {
  test('renders no products message when data is empty', () => {
    render(<ProductList data={[]} />);

    expect(screen.getByText('No products in list')).toBeInTheDocument();
  });

  test('should match snapshot for ProductCard', () => {
    const { container } = render(<ProductList data={[]} />);

    expect(container).toMatchSnapshot();
  });

  test('should render ProductList successfully', () => {
    render(<ProductList data={MOCK_PRODUCT_LIST} />);

    expect(screen.getByText(MOCK_PRODUCT_LIST[0].name)).toBeInTheDocument();
    expect(screen.getByText(MOCK_PRODUCT_LIST[1].name)).toBeInTheDocument();
  });
});
