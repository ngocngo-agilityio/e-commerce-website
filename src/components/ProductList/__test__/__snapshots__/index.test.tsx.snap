// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList component should match snapshot for ProductCard 1`] = `
<div>
  <p
    class="chakra-text css-xi606m"
  >
    No products in list
  </p>
</div>
`;
